File: src/euler3.c
/*
 * Problem 3 - Largest prime factor
 *
 * The prime factors of 13195 are 5, 7, 13 and 29.
 * What is the largest prime factor of the number 600851475143 ?
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int isprime(long int n) {
    long int stop = floor(sqrt(n));
    for (long int i = 2; i <= stop; i++){
        if (n%i == 0) {
            return 0;
        }
    }
    return 1;
}

int trialdiv(long int n) {
    long int curr = n;
    long int stop = floor(sqrt(n));
    long int primes [stop];
    long int idx = 0;
    for (long int i = 2; i <= stop; i++) {
        if (curr%i == 0) {
            curr = curr/i;
            primes[idx] = i;
            idx += 1;
        }
    }
    primes[idx] = curr;
    for (long int i = 0; i <= idx; i++) {
        if (isprime(primes[i]) == 1) {
            printf("%ld\n", primes[i]);
        }
    }
    return 0;
}

int main(int argc, char **argv) {
    long int n = atoi(argv[1]);
    // printf("%d", isprime(n));
    printf("Prime factors of %ld are:\n", n);
    trialdiv(n);
    return 0;
}

File: python/euler3.py
#!/usr/bin/env python3

import numpy as np

"""
Problem 3 - Largest prime factor

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600,851,475,143?
"""

def isprime(x):
    for i in range(2, int(np.floor(x**0.5))+1):
        if x%i == 0:
            return False
    return True

def prime_sieve(n):
    primes = list(range(2, n+1))
    for i, val in enumerate(primes):
        if primes[i] == 0:
            continue
        else:
            j = i + val
            while j < len(primes):
                primes[j] = 0
                j += val
    primes = [p for p in primes if p]
    return primes

def prime_factor(n=13195):
    factors = []
    curr = n
    for i in range(2, int(np.floor(n**0.5))+1):
        if curr%i == 0:
            factors.append(i)
            curr /= i
    factors.append(int(curr))
    factors = [i for i in factors if isprime(i)]
    return factors

class Test(object):
    def __init__(self):
        self.length = 10

    def __str__(self):
        return str(self.length)

if __name__ == '__main__':
    a = prime_factor(600851475143)
#     b = prime_factor(2*4*6*8*3*5)
#     c = prime_factor(17)
    print(a)
#     print(b)
#     print(c)

    # print('hello')
    test = Test()
    print(str(test))
